lib/node.js

```diff
@@ -299,6 +310,90 @@ class Node {
     this.edgesIn.forEach(edge => edge.reload(true))
   }
 
+  // node.explain(nodes seen already, edge we're trying to satisfy
+  // if edge is not specified, it lists every edge into the node.
+  explain (edge = null, seen = []) {
+    if (this[_explanation])
+      return this[_explanation]
+
+    return this[_explanation] = this[_explain](edge, seen)
+  }
+
+  [_explain] (edge, seen) {
+    if (this.isRoot && !this.sourceReference) {
+      return {
+        location: this.path
+      }
+    }
+
+    const why = {
+      name: this.isRoot ? this.package.name : this.name,
+      version: this.package.version,
+    }
+    if (this.errors.length || !this.package.name && !this.package.version) {
+      why.errors = this.errors.length ? this.errors : [
+        new Error('invalid package: lacks name and/or version')
+      ]
+      why.package = this.package
+    }
+
+    if (this.root.sourceReference) {
+      const {name, version} = this.root.package
```

Comment:
Is there value in using separated `name` and `version` instead of `pkgid` for `whileInstalling`?

Reply:
Occasionally the `node.package` won't have that, if it's being set by something other than reading a `package.json`, I believe?  I seem to recall that showing up in some cases I was poking at, but didn't dig into it.